tensorboard/webapp/webapp_data_source/tb_feature_flag_data_source.ts

```diff
@@ -40,62 +34,39 @@ export class QueryParamsFeatureFlagDataSource
-    const featureFlags: Partial<FeatureFlags> = enableMediaQuery
-      ? this.getPartialFeaturesFromMediaQuery()
-      : {};
-    if (params.has(EXPERIMENTAL_PLUGIN_QUERY_PARAM_KEY)) {
-      featureFlags.enabledExperimentalPlugins = params.getAll(
-        EXPERIMENTAL_PLUGIN_QUERY_PARAM_KEY
-      );
-    }
-    if (params.has('tensorboardColab')) {
-      featureFlags.inColab = params.get('tensorboardColab') === 'true';
-    }
-    if (params.has(SCALARS_BATCH_SIZE_PARAM_KEY)) {
-      featureFlags.scalarsBatchSize = Number(
-        params.get(SCALARS_BATCH_SIZE_PARAM_KEY)
-      );
-    }
-
-    if (params.has(ENABLE_COLOR_GROUP_QUERY_PARAM_KEY)) {
-      featureFlags.enabledColorGroup =
-        params.get(ENABLE_COLOR_GROUP_QUERY_PARAM_KEY) !== 'false';
-    }
-
-    if (params.has(ENABLE_COLOR_GROUP_BY_REGEX_QUERY_PARAM_KEY)) {
-      featureFlags.enabledColorGroupByRegex =
-        params.get(ENABLE_COLOR_GROUP_BY_REGEX_QUERY_PARAM_KEY) !== 'false';
-    }
-
-    if (params.has(ENABLE_DARK_MODE_QUERY_PARAM_KEY)) {
-      featureFlags.defaultEnableDarkMode =
-        params.get(ENABLE_DARK_MODE_QUERY_PARAM_KEY) !== 'false';
-    }
-
-    if (params.has(ENABLE_LINKED_TIME_PARAM_KEY)) {
-      featureFlags.enabledLinkedTime =
-        params.get(ENABLE_LINKED_TIME_PARAM_KEY) !== 'false';
-    }
-
-    if (params.has(ENABLE_CARD_WIDTH_SETTING_PARAM_KEY)) {
-      featureFlags.enabledCardWidthSetting =
-        params.get(ENABLE_CARD_WIDTH_SETTING_PARAM_KEY) !== 'false';
-    }
+    const featureFlags: Partial<Record<keyof FeatureFlags, FeatureFlagType>> =
+      enableMediaQuery ? this.getPartialFeaturesFromMediaQuery() : {};
+    Object.entries(FeatureFlagMetadataMap).forEach(
+      ([flagName, flagMetadata]) => {
+        const featureValue = this.getFeatureValue(
+          flagMetadata as FeatureFlagMetadata<boolean>
+        );
+        if (featureValue !== null) {
+          const f = flagName as keyof FeatureFlags;
+          featureFlags[f] = featureValue;
+        }
+      }
+    );
+    return featureFlags as Partial<FeatureFlags>;
+  }
 
-    if (params.has(FORCE_SVG_RENDERER)) {
-      featureFlags.forceSvg = params.get(FORCE_SVG_RENDERER) !== 'false';
+  protected getFeatureValue<T>(flagMetadata: FeatureFlagMetadata<T>) {
+    const params = this.queryParams.getParams();
+    const queryParamOverride = flagMetadata.queryParamOverride;
+    if (!queryParamOverride || !params.has(queryParamOverride)) {
+      return null;
     }
-
-    if (params.has(ENABLE_DATA_TABLE_PARAM_KEY)) {
-      featureFlags.enabledScalarDataTable =
-        params.get(ENABLE_DATA_TABLE_PARAM_KEY) !== 'false';
+    const paramValues: T[] = params.getAll(queryParamOverride).map((value) => {
+      return flagMetadata.parseValue(value);
+    });
+    if (!paramValues.length) {
+      return null;
     }
 
-    return featureFlags;
+    return flagMetadata.isArray ? paramValues.flat() : paramValues[0];
```

Comment:
Could be worth describing what case you are trying to support with isArray and flat() here, similarly to how you documented the use case where you implemented the inverse functionality.